embassy-stm32/src/usart/ringbuffered.rs

```diff
@@ -210,28 +170,11 @@ impl<'d, T: BasicInstance, RxDma: super::RxDma<T>> RingBufferedUartRx<'d, T, RxD
 
             compiler_fence(Ordering::SeqCst);
 
-            // SAFETY: read only and we only use Rx related flags
-            let sr = unsafe { sr(r).read() };
-
-            // SAFETY: only clears Rx related flags
-            unsafe {
-                // This read also clears the error and idle interrupt flags on v1.
-                rdr(r).read_volatile();
-                clear_interrupt_flags(r, sr);
-            }
+            // Critical section is needed so that IDLE isn't set after
+            // our read but before we clear it.
+            let sr = critical_section::with(|_| clear_idle_flag(T::regs()));
```

Comment:
Instead of having this here in a critical section, could we put it directly in the usart on_interrupt? and then from there capture the SR register in an AtomicU32 that can be safely swapped and analyzed here?

Reply:
Register access is fast, atomics are slow. However, the critical section may be slower. I personally don't like sharing state between interrupt handlers, though.